onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc

```diff
@@ -1,139 +1,277 @@
+  bool HasGraphOutput() const noexcept { return !dst.has_value(); }
+
+  const std::optional<NodeInfo>& GetNodeInfoAtEnd(End end) const {
+    return end == End::Source ? src : dst;
+  }
+
+  Node* GetNodeAtEnd(Graph& graph, End end) const {
+    if (const auto& node_info = GetNodeInfoAtEnd(end); node_info.has_value()) {
+      Node* node = graph.GetNode(node_info->node_idx);
+      ORT_ENFORCE(node != nullptr, "Invalid node index ", node_info->node_idx);
+      return node;
+    }
+    return nullptr;
+  }
+
+  const Node* GetNodeAtEnd(const Graph& graph, End end) const {
+    if (const auto& node_info = GetNodeInfoAtEnd(end); node_info.has_value()) {
+      const Node* node = graph.GetNode(node_info->node_idx);
+      ORT_ENFORCE(node != nullptr, "Invalid node index ", node_info->node_idx);
+      return node;
+    }
+    return nullptr;
+  }
+
+  static ExtendedGraphEdge FromGraphEdge(const graph_utils::GraphEdge& graph_edge) {
+    return ExtendedGraphEdge{
+        NodeInfo{graph_edge.src_node, graph_edge.src_arg_index},
+        NodeInfo{graph_edge.dst_node, graph_edge.dst_arg_index},
+        graph_edge.arg_name};
+  }
+
+  static std::optional<ExtendedGraphEdge> FromInputOrInitializerToNode(
+      const Graph& graph, const Node& node, int node_input_def_idx) {
+    const auto node_inputs = node.InputDefs();
+    ORT_ENFORCE(node_input_def_idx >= 0 &&
+                static_cast<size_t>(node_input_def_idx) < node_inputs.size());
+
+    const auto* node_input = node_inputs[node_input_def_idx];
+    if (!graph.IsInputsIncludingInitializers(node_input)) {
+      return std::nullopt;
+    }
 
-static bool CanNodePropagate(const Node& node) {
+    return ExtendedGraphEdge{
+        std::nullopt,
+        NodeInfo{node.Index(), node_input_def_idx},
+        node_input->Name()};
+  }
+
+  static std::optional<ExtendedGraphEdge> FromNodeToOutput(
+      const Graph& graph, const Node& node, int node_output_def_idx) {
+    const auto node_outputs = node.OutputDefs();
+    ORT_ENFORCE(node_output_def_idx >= 0 &&
+                static_cast<size_t>(node_output_def_idx) < node_outputs.size());
+
+    const auto* node_output = node_outputs[node_output_def_idx];
+    if (!graph.IsOutput(node_output)) {
+      return std::nullopt;
+    }
+
+    return ExtendedGraphEdge{
+        NodeInfo{node.Index(), node_output_def_idx},
+        std::nullopt,
+        node_output->Name()};
+  }
+
+  // there is also the case where the graph input is an output, but we don't care about that for this transformer
+};
+
+bool CanNodePropagate(const Node& node) {
   return graph_utils::IsSupportedOptypeVersionAndDomain(node, "MaxPool", {12}) ||
          graph_utils::IsSupportedOptypeVersionAndDomain(node, "Reshape", {5, 13, 14}) ||
          graph_utils::IsSupportedOptypeVersionAndDomain(node, "Transpose", {1, 13});
 }
 
-static bool TryCancelOutDQQPair(Graph& graph, Node& dq_node, Node& q_node) {
```

Comment:
@yufenglee This change removes the DQ/Q cancellation. This transformer shouldn't create any such pairs, but matching DQ/Q pairs in the original model will no longer be removed. Is this fine?

Reply:
yes, it is good as we won't insert QDQ if the tensor is already quantized.